File: trunk/main/src/com/seovic/lang/Defaults.java
/*
 * Copyright 2009 Aleksandar Seovic
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.seovic.lang;


import java.lang.reflect.Constructor;

import java.util.Properties;

import java.io.IOException;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;


/**
 * Factory class for default implementations.
 *
 * @author Aleksandar Seovic  2009.09.28
 */
public class Defaults
    {
    /**
     * Singleton constructor.
     */
    @SuppressWarnings("unchecked")
    private Defaults()
        {
        Properties props = loadDefaults();
        m_ctorExpression = getConstructor(props.getProperty(EXPRESSION_TYPE));
        m_ctorExtractor  = getConstructor(props.getProperty(EXTRACTOR_TYPE));
        m_ctorUpdater    = getConstructor(props.getProperty(UPDATER_TYPE));
        m_ctorCondition  = getConstructor(props.getProperty(CONDITION_TYPE));
        m_scriptLanguage = props.getProperty(SCRIPT_LANGUAGE);
        }

    // ---- public methods --------------------------------------------------

    /**
     * Creates a default expression instance.
     *
     * @param expression  string expression
     *
     * @return expression instance
     */
    public static Expression createExpression(String expression)
        {
        try
            {
            return instance.m_ctorExpression.newInstance(expression);
            }
        catch (Exception e)
            {
            throw new RuntimeException(e);
            }
        }

    /**
     * Creates a default extractor instance.
     *
     * @param expression  string expression
     *
     * @return extractor instance
     */
    public static Extractor createExtractor(String expression)
        {
        try
            {
            return instance.m_ctorExtractor.newInstance(expression);
            }
        catch (Exception e)
            {
            throw new RuntimeException(e);
            }
        }

    /**
     * Creates a default updater instance.
     *
     * @param expression  string expression
     *
     * @return updater instance
     */
    public static Updater createUpdater(String expression)
        {
        try
            {
            return instance.m_ctorUpdater.newInstance(expression);
            }
        catch (Exception e)
            {
            throw new RuntimeException(e);
            }
        }

    /**
     * Creates a default condition instance.
     *
     * @param expression  conditional expression
     *
     * @return condition instance
     */
    public static Condition createCondition(String expression)
        {
        try
            {
            return instance.m_ctorCondition.newInstance(expression);
            }
        catch (Exception e)
            {
            throw new RuntimeException(e);
            }
        }

    /**
     * Return the name of the default script language.
     *
     * @return the name of the default script language
     */
    public static String getScriptLanguage()
        {
        return instance.m_scriptLanguage;
        }


    // ---- helper methods --------------------------------------------------

    /**
     * Loads default values from a com/seovic/lang/defaults.properties file.
     *
     * @return loaded Properties instance
     */
    protected Properties loadDefaults()
        {
        Properties props = new Properties();
        try
            {
            props.load(Defaults.class.getResourceAsStream("defaults.properties"));
            }
        catch (IOException e)
            {
            // should never happen
            props.setProperty(EXPRESSION_TYPE, DEFAULT_EXPRESSION_TYPE);
            props.setProperty(EXTRACTOR_TYPE,  DEFAULT_EXTRACTOR_TYPE);
            props.setProperty(UPDATER_TYPE,    DEFAULT_UPDATER_TYPE);
            props.setProperty(CONDITION_TYPE,  DEFAULT_CONDITION_TYPE);
            props.setProperty(SCRIPT_LANGUAGE, DEFAULT_SCRIPT_LANGUAGE);

            log.warn("Configuration resource com/seovic/lang/defaults.properties"
                     + " not found. Using hardcoded defaults: " + props);
            }
        return props;
        }

    /**
     * Gets a constructor from the specified class that accepts a single String
     * argument.
     *
     * @param className  the name of the class to find the constructor in
     *
     * @return a constructor for the specified class that accepts a single
     *         String argument
     */
    protected Constructor getConstructor(String className)
        {
        try
            {
            Class cls = Class.forName(className);
            return cls.getConstructor(String.class);
            }
        catch (Exception e)
            {
            throw new RuntimeException("Unable to find a constructor that accepts"
                                       + " a single String argument in the "
                                       + className + " class.", e);
            }
        }


    // ---- constants -------------------------------------------------------

    /**
     * Property names
     */
    private static final String EXPRESSION_TYPE = "expression.type";
    private static final String EXTRACTOR_TYPE  = "extractor.type";
    private static final String UPDATER_TYPE    = "updater.type";
    private static final String CONDITION_TYPE  = "condition.type";
    private static final String SCRIPT_LANGUAGE = "script.language";

    /**
     * Default property values
     */
    private static final String DEFAULT_EXPRESSION_TYPE = "com.seovic.lang.expression.MvelExpression";
    private static final String DEFAULT_EXTRACTOR_TYPE  = "com.seovic.lang.extractor.ExpressionExtractor";
    private static final String DEFAULT_UPDATER_TYPE    = "com.seovic.lang.updater.ExpressionUpdater";
    private static final String DEFAULT_CONDITION_TYPE  = "com.seovic.lang.condition.ExpressionCondition";
    private static final String DEFAULT_SCRIPT_LANGUAGE = "javascript";


    // ---- static members --------------------------------------------------

    /**
     * Singleton instance.
     */
    private static final Defaults instance = new Defaults();

    /**
     * Logger instance.
     */
    private static final Log log = LogFactory.getLog(Defaults.class);


    // ---- data members ----------------------------------------------------

    /**
     * Constructor for default expression type.
     */
    private Constructor<Expression> m_ctorExpression;

    /**
     * Constructor for default extractor type.
     */
    private Constructor<Extractor> m_ctorExtractor;

    /**
     * Constructor for default updater type.
     */
    private Constructor<Updater> m_ctorUpdater;

    /**
     * Constructor for default condition type.
     */
    private Constructor<Condition> m_ctorCondition;

    /**
     * The name of the default script language.
     */
    private String m_scriptLanguage;
    }

File: main.net/src/Coherence.Commons/Core/Defaults.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Configuration;
using System.Reflection;

namespace Seovic.Coherence.Core
{
    public class Defaults
    {
        #region Constructors

        /// <summary>
        /// Singleton constructor.
        /// </summary>
        private Defaults()
        {
            IDictionary<string, string> dict = LoadDefaults();
            m_ctorExpression = GetConstructor(dict[EXPRESSION_TYPE]);
            m_ctorExtractor  = GetConstructor(dict[EXTRACTOR_TYPE]);
            m_ctorUpdater    = GetConstructor(dict[UPDATER_TYPE]);
            m_ctorCondition  = GetConstructor(dict[CONDITION_TYPE]);
            m_scriptLanguage = dict[SCRIPT_LANGUAGE];
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Creates a default expression instance.
        /// </summary>
        /// <param name="expression">Stirng expression.</param>
        /// <returns>Expression intance</returns>
        public static IExpression CreateExpression(string expression)
        {
            return (IExpression) Instance.m_ctorExpression.Invoke(new[] {expression});
        }

        /// <summary>
        /// Creates a default extractor instance.
        /// </summary>
        /// <param name="expression">String expression</param>
        /// <returns>Extractor instance</returns>
        public static IExtractor CreateExtractor(string expression)
        {
            return (IExtractor) Instance.m_ctorExtractor.Invoke(new[] {expression});
        }

        /// <summary>
        /// Creates a default updater instance.
        /// </summary>
        /// <param name="expression">String expression.</param>
        /// <returns>Updater instance.</returns>
        public static IUpdater CreateUpdater(string expression)
        {
            return (IUpdater) Instance.m_ctorUpdater.Invoke(new[] {expression});
        }

        /// <summary>
        /// Creates a default condition instance.
        /// </summary>
        /// <param name="expression">Conditional expression</param>
        /// <returns>Condition instance</returns>
        public static ICondition CreateCondition(String expression)
        {
            return (ICondition) Instance.m_ctorCondition.Invoke(new[] {expression});
        }
    
        /// <summary>
        /// Return the name of the default script language.
        /// </summary>
        /// <returns>
        /// The name of the default script language.
        /// </returns>
        public static string GetScriptLanguage()
        {
            return Instance.m_scriptLanguage;
        }

        #endregion

        #region Helper methods

        /// <summary>
        /// Loads default values from a application configuration file.
        /// </summary>
        /// <returns>
        /// Dictinary containing configuration values.
        /// </returns>
        protected IDictionary<string, string> LoadDefaults()
        {
            IDictionary<string, string> props = new Dictionary<string, string>();
            try
            {
                NameValueCollection config = ConfigurationManager.AppSettings;
                foreach (string key in config.Keys)
                {
                    props.Add(key, config[key]);
                }
            }
            catch (Exception e)
            {
                // should never happen
                props.Add(EXPRESSION_TYPE, DEFAULT_EXPRESSION_TYPE);
                props.Add(EXTRACTOR_TYPE,  DEFAULT_EXTRACTOR_TYPE);
                props.Add(UPDATER_TYPE,    DEFAULT_UPDATER_TYPE);
                props.Add(CONDITION_TYPE,  DEFAULT_CONDITION_TYPE);
                props.Add(SCRIPT_LANGUAGE, DEFAULT_SCRIPT_LANGUAGE);
                // TODO: log
                //log.warn("Configuration resource com/seovic/lang/defaults.properties"
                //         + " not found. Using hardcoded defaults: " + props);
            }
            return props;
        }

        /// <summary>
        /// Gets a constructor from the specified class that accepts a 
        /// single String argument.
        /// </summary>
        /// <param name="typeName">
        /// The name of the type to find the constructor in
        /// </param>
        /// <returns>
        /// A constructor for the specified class that accepts a
        /// single String argument
        /// </returns>
        protected ConstructorInfo GetConstructor(string typeName)
        {
            Type cls = Type.GetType(typeName);
            return cls.GetConstructor(new[] {typeof(String)});
        }

        #endregion

        #region Constants
        
        #region Configuration property keys

        private static readonly string EXPRESSION_TYPE = "expression.type";
        private static readonly string EXTRACTOR_TYPE  = "extractor.type";
        private static readonly string UPDATER_TYPE    = "updater.type";
        private static readonly string CONDITION_TYPE  = "condition.type";
        private static readonly string SCRIPT_LANGUAGE = "script.language";

        #endregion

        #region Configuration default values

        // TBD
        private static readonly string DEFAULT_EXPRESSION_TYPE = "Seovic.Coherence.Core.Expression.MvelExpression, Coherence.Commons";
        private static readonly string DEFAULT_EXTRACTOR_TYPE  = "Seovic.Coherence.Core.Extractor.ExpressionExtractor, Coherence.Commons";
        private static readonly string DEFAULT_UPDATER_TYPE    = "Seovic.Coherence.Core.Updater.ExpressionUpdater, Coherence.Commons";
        private static readonly string DEFAULT_CONDITION_TYPE  = "Seovic.Coherence.Core.Condition.ExpressionCondition, Coherence.Commons";
        private static readonly string DEFAULT_SCRIPT_LANGUAGE = "javascript";
        
        #endregion

        #endregion

        #region Static members

        private static readonly Defaults Instance = new Defaults();

        #endregion

        #region Data members

        /// <summary>
        /// Constructor for default expression type.
        /// </summary>
        private ConstructorInfo m_ctorExpression;

        /// <summary>
        /// Constructor for default extractor type.
        /// </summary>
        private ConstructorInfo m_ctorExtractor;

        /// <summary>
        /// Constructor for default updater type.
        /// </summary>
        private ConstructorInfo m_ctorUpdater;

        /// <summary>
        /// Constructor for default condition type.
        /// </summary>
        private ConstructorInfo m_ctorCondition;

        /// <summary>
        /// The name of the default script language.
        /// </summary>
        private String m_scriptLanguage;

        #endregion

    }
}
